### DISCUSSION TITLE
output slide file format

### DISCUSSION THREAD

**Discussion by NicolasNerr** [MathOnco/valis#21] [General]
Hello ! 

Thank you for the tool, the results are great ! 

I am actually using a slide viewer that allows us to post some of our algorithms results and visualize them directly on the slide. 
However, one of the format that is not readable is the ome.tiff format. 

Do you plan to have multiple output format ? 

Another related question is which python library you are using to read and write ome.tiff slides ? so maybe I can further improve our plateform as it is now nearly entirely based on openslide and its supported formats.

Thank you so much, 
Best regards.

**Comment by cdgatenbee** [comment]
Hi @NicolasNerr,

So glad to hear valis is working for you! Valis uses Openslide, Bio-formats, and pyvips to read the images, and you can use valis to read the images with the appropriate package by doing: 

```
reader_cls = slide_io.get_slide_reader(src_f)
reader = reader_cls(src_f)
```

This ```reader``` will then be able to convert the slide to either a numpy array (```reader.slide2image(level)```) or a pyvips.Image (```reader.slide2vips(level)```), where ```level``` is the pyramid level you would like to read. It also extracts some metadata, like pyramid level sizes, resolution, channel names, etc... More info can be found in the [slide_io docs](https://valis.readthedocs.io/en/latest/slide_io.html).

 pyvips is used to write the images, so it is possible to save the registered images in a variety of formats (check out the [pyvips.Image](https://libvips.github.io/pyvips/vimage.html) documentation). Right now, valis only writes ome.tiff since it's the only open-source format that saves the WSI image and its metadata. However, you could also save each image as a pyramid tiff and then load it using your slide viewer. You could also try adding the ome-xml or original metadata to the tiff's image description, but I'm not sure how the slide viewer would handle it. I could imagine it might still work just fine, but it could also be ignored or throw an error. Hard to say without trying. If you'd like to take this approach, let me know and I can put together some example code on how you would warp each slide and then save as a pyramid tiff using pyvips.

Best,
Chandler
